### PR TITLE
11956 diophantine cornacchia

### LINKED ISSUE #11956: diophantine doesn't find solutions for 2*x**2+y**2-16
Diophantine returns the empty set but `{x=0,y=4}` and `{x=0,y=-4}` are solutions.
I suspect there is some issue with the solving of the elliptic case, as for example `10*x**2 + 12*x*y + 12*y**2 - 34` also doesn't return all solutions: it returns `{x=-1,y=2}` and `{x=1,y=1}` but fails to find `{x=-1,y=-1}` and `{x=1,y=-2}`

### PR DESCRIPTION
**Pull request by pranjaltale16**
Fixes #11956 #11959 

Corrected `diophantene()` function mainly modified 'cornacchia()' function
Before diophantine solution for the eqn `x**2+2*y**2 -32` is null set but now it is giving correct solutions
Also initially in cornacchia function when 'a = b = 1' it was returning `set([(3, 4)])` instead of `set([(4, 3)])`
And modified  `diop_DN()` function to show all solutions rather than half of them.

### CODE REVIEW

**Comment by pranjaltale16** [comment]
There is property of `cornacchia()` function to return primitive solutions plus when `a=b=1` then it returns the solution with `x>=y` only while in tests of diophantine 
`Failed example:`
`    cornacchia(1, 1, 25) # equation x**2 + y**2 = 25`
`Expected:`
`    set([(3, 4)])`
`Got:`
`    set([(4, 3)])`
Actually it should get `set([4, 3])` only so i made changes in test for that mainly.

Also for `diop_DN` this was returning half of the solutions but now it is returning the complete set for ex
`diop_solve(x**2 + y**2 - 5)
 set([(-2, -1), (-2, 1), (-1, -2), (-1, 2), (1, -2), (1, 2), (2, -1), (2, 1)])`
intially it was returning
`set([(-1, -2), (-1, 2), (1, -2), (1, 2)])` only

References
  http://docs.sympy.org/latest/modules/solvers/diophantine.html#cornacchia
  http://www.numbertheory.org/php/cornacchia.html

**Comment by asmeurer** [comment]
@smichr @thilinarmtb

**Comment by thilinarmtb** [comment]
@pranjaltale16 : Thanks for the PR. I am sorry I have forgotten the properties of cornacchia function. Can you please update me on why we are having such a property? Is it documented somewhere?

**Comment by pranjaltale16** [comment]
@thilinarmtb

**Comment by pranjaltale16** [comment]
@thilinarmtb 
I made changes at two places in solvers.diophantine.py

1. Changes in `cornacchia` funtion with reference to 
    http://docs.sympy.org/latest/modules/solvers/diophantine.html#cornacchia and 
    http://www.numbertheory.org/php/cornacchia.html
The issue is `cornacchia(1,1,25)` should return `set([4,3])` but here it is returning `set([3,4])`. So i made changes for that also for `cornacchia(a,b,m)` if `a =b =1` then it should reuturn the solution with `x>=y` while it is also returning the solution corresponding to `x<y` ex:(3,4) for cornacchia(1,1,25).

2. Changes in `diop_DN` function this is with ref to #11959 
    what happening here is `diop_DN(1,-D,N)` calls the `cornacchia` function and when `D = -1 ` it returns only one set of solution with `x>=y`, if our eqx has been reduced in the form of `x**2 +y**2 = m` that is `D = -1` then if `(x,y)` is the solution with `x>=y` returned from `cornacchia` then `(y,x)` should also be the solution which can't be returned from `cornacchia` so i made changes in `diop_DN` so that when D = -1 it considers (y, x) also for solution.
    Ref: https://www.alpertron.com.ar/QUAD.HTM 
I considered the equation `10*x**2 - 6*x + 10*y**2 - 14*y - 52548` mentioned in #11959 the original code is returning only 8 solutions while it will have 16 checked from the Ref

After making changes both the above problems have been resolved

**Comment by pranjaltale16** [comment]
ping, @thilinarmtb @smichr 
Please look into this -  @dharasty made  a file with 162 cases corresponding to `diophantine` when i  tested it with "Sympy1.0" i got 137 cases failed out of 162, and after testing it with code corrections i made it passed all the 162 cases.
For the test case file you can find it at #11959

**Comment by pranjaltale16** [comment]
Ping @thilinarmtb @smichr - -  please look into this do i have to work or update anything..? 
Any issue in this?

**Comment by jksuom** [comment]
The changes look good to me, but the tests in test_diophantine.py should be fixed to include the new results. (Run 'bin/test diophantine' to see what should be changed.) Also SymPy code is expected have a space on either side of comparison operators (==, <, ...) and after commas. Please add those spaces.

**Comment by pranjaltale16** [comment]
@jksuom 
I added spaces for comparisions and commas also and update the.

Also I have already made changes for tests in `test_diophantine.py` [commit](https://github.com/sympy/sympy/pull/11961/commits/41ef5dc175d2d6287677d1c31b07895bb27c875e) , and in my local machine all the tests passed.

`executable:         /home/pranjal/anaconda2/bin/python  (2.7.12-final-0) [CPython]`
``architecture:       64-bit``
`cache:              yes`
`ground types:       python` 
`random seed:        19786098`
`hash randomization: on (PYTHONHASHSEED=2763570774)`

`sympy/solvers/tests/test_diophantine.py[42] ......f..w................................                  [OK]`
`tests finished: 40 passed, 1 skipped, 1 expected to fail, in 16.73 seconds`

File: sympy/solvers/diophantine.py [hunk 1]
```diff
@@ -1301,7 +1303,7 @@ def cornacchia(a, b, m):
     >>> cornacchia(2, 3, 35) # equation 2x**2 + 3y**2 = 35
     set([(2, 3), (4, 1)])
     >>> cornacchia(1, 1, 25) # equation x**2 + y**2 = 25
-    set([(3, 4)])
+    set([(4, 5)])
```
Comment (jksuom) [inline sympy/solvers/diophantine.py]: Should this be (4, 3)?

File: sympy/solvers/diophantine.py [hunk 2]
```diff
@@ -1323,11 +1325,13 @@ def cornacchia(a, b, m):
         return None
 
     for t in v:
-        if t < m // 2:
+        if t >= m // 2:
+            u, r = t, m
+        elif t < m // 2:
+            u,r = m-t,m
```
Comment (jksuom) [inline sympy/solvers/diophantine.py]: Please add a space after commas.

File: sympy/solvers/tests/test_diophantine.py [hunk 3]
```diff
@@ -199,8 +199,8 @@ def test_DN():
 
     # When equation is x**2 + y**2 = N
     # Solutions are interchangeable
-    assert diop_DN(-1, 5) == [(1, 2)]
-    assert diop_DN(-1, 169) == [(5, 12), (0, 13)]
+    assert diop_DN(-1, 5) == [(2, 1),(1, 2)]
```
Comment (jksuom) [inline sympy/solvers/tests/test_diophantine.py]: Spaces after commas on these lines too.

File: sympy/solvers/diophantine.py (hunk 1 above)
Comment (pranjaltale16) [inline sympy/solvers/diophantine.py]: @jksuom extremely sorry for this.

File: sympy/solvers/diophantine.py (hunk 2 above)
Comment (pranjaltale16) [inline sympy/solvers/diophantine.py]: Done

File: sympy/solvers/tests/test_diophantine.py (hunk 3 above)
Comment (pranjaltale16) [inline sympy/solvers/tests/test_diophantine.py]: Done

**Comment by pranjaltale16** [comment]
@jksuom 
Sorry for such silly mistakes.
I have made all the changes suggested.

**Comment by jksuom** [comment]
I forgot to point out that also bin/doctest should be run. In the docstring of diop_solve, there is an example whose result was changed: 
`>>> diop_solve(x**2 + y**2 - 5)`
There are now four solutions instead of two.

It also seems that there is another test that was added to test_diopcoverage after your branch was checked out:
`assert diop_quadratic(x**2 + y**2 - 1**2 - 3**4) == set([(-1, 9), (1, 9)])`.

To fix that you should probably synchronize with master and then add the missing results.

**Comment by pranjaltale16** [comment]
@jksuom 
`diop_solve(x**2 + y**2 - 5 )` should have 8 solutions ?? 
`set([(-2, -1), (-2, 1), (-1, -2), (-1, 2), (1, -2), (1, 2), (2, -1), (2, 1)])`

Also after merging master to the branch, the solutions are not coming proper for diophantine
Before merging master all the tests are running fine.

I updated the docstring for diop_solve but facing issues in merging master to my local branch.

**Comment by jksuom** [comment]
> There are now four solutions instead of two.

I was probably thinking of only those where |x| > |y|. You should add all solutions, of course.

> Before merging master all the tests are running fine.

That is true. But travis seems to include the new test with `diop_quadratic(x**2 + y**2 - 1**2 - 3**4)`.
It has to be included (in amended form) to make the tests pass. Otherwise this cannot be merged.

If you cannot merge master into your branch, you could consider cherry-picking the commit 8c003e8d1f0 introducing the test and then editing the test.

**Comment by pranjaltale16** [comment]
@jksuom 
Could you please help me out. 
I am struck in this. I tried to use cherry-pick method also still i am unable to do so. is there any other approach.
Also I tried to merge master first and then apply changes in a different branch but still I am getting wrong answers.

**Comment by jksuom** [comment]
It looks like the merge did succeed, and I also removed some whitespace at the end of the new example line.

**Comment by jksuom** [comment]
```In [3]: diop_solve(x**2 + y**2 - 5)
Out[3]: set([(-2, 1), (-1, 2), (1, 2), (2, 1)])
```
This is the result of the current code. Did you plan to include the solutions with all sign variations?

**Comment by pranjaltale16** [comment]
Yes, whenever we find solution it should give us all the solutions.

**Comment by jksuom** [comment]
Then something is missing. Since I edited your github repo, you should probably synchronize your local branch with the github branch before continuing.

**Comment by pranjaltale16** [comment]
Sure, I am also facing issues in merging master.
I will look into it.

**Comment by jksuom** [comment]
Your github branch is now merged with master. You should pull that to your local branch. If there are problems, in might be possible to rename the local branch and clone the github branch on the local machine (with the original name).

**Comment by pranjaltale16** [comment]
@jksuom I was facing several difficulties in merging, so I created a new branched merged master in it first and then made all the changes to it. 
All the tests and doctest passed for that also i chekced with 162 cases for diophantine they all passed too.

I made a new pull request -- #11989, Please look into it.

**Pull request closed without merging**